src/Compilers/Core/Portable/Operations/OperationFactory.cs

```diff
@@ -1,29 +1,42 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System.Collections.Immutable;
-using System.Linq;
 
 namespace Microsoft.CodeAnalysis.Semantics
 {
     internal static class OperationFactory
     {
-        public static IVariableDeclaration CreateVariableDeclaration(ILocalSymbol variable, IOperation initialValue, SemanticModel semanticModel, SyntaxNode syntax)
+        public static IVariableDeclaration CreateVariableDeclaration<TEqualsValueSyntax>(ILocalSymbol variable, IOperation initialValue, SemanticModel semanticModel, SyntaxNode syntax)
+            where TEqualsValueSyntax : SyntaxNode
         {
-            return CreateVariableDeclaration(ImmutableArray.Create(variable), initialValue, semanticModel, syntax);
+            return CreateVariableDeclaration<TEqualsValueSyntax>(ImmutableArray.Create(variable), initialValue, semanticModel, syntax);
         }
 
-        public static VariableDeclaration CreateVariableDeclaration(ImmutableArray<ILocalSymbol> variables, IOperation initialValue, SemanticModel semanticModel, SyntaxNode syntax)
+        public static VariableDeclaration CreateVariableDeclaration<TEqualsValueSyntax>(ImmutableArray<ILocalSymbol> variables, IOperation initialValue, SemanticModel semanticModel, SyntaxNode syntax)
+            where TEqualsValueSyntax : SyntaxNode
         {
             return new VariableDeclaration(
                 variables,
-                initialValue,
+                CreateLocalInitializer<TEqualsValueSyntax>(initialValue, semanticModel),
                 semanticModel,
                 syntax,
                 type: null,
                 constantValue: default(Optional<object>),
                 isImplicit: false); // variable declaration is always explicit
         }
 
+        private static ILocalInitializer CreateLocalInitializer<TEqualsValueSyntax>(IOperation initialValue, SemanticModel semanticModel)
+            where TEqualsValueSyntax : SyntaxNode
+        {
+            if (initialValue == null)
+            {
+                return null;
+            }
+
+            var syntax = initialValue.Syntax.Parent as TEqualsValueSyntax ?? initialValue.Syntax;
```

Comment:
When do we fallback to `initialValue.Syntax`?

Reply:
This is primarily for VB's `AsNewClauseSyntax`, and also I wasn't sure if error cases can have a different parent syntax node.

Reply:
Consider moving the `syntax` calculation to the caller of `CreateVariableDeclaration`. It looks like there is only one caller in C# and one in VB where there may be an initializer. That removes the magic of walking up to the parent in the shared code and would remove the generic type parameter.

Reply:
Perhaps move the entire call to `CreateLocalInitializer` to the caller of `CreateVariableDeclaration`.

Reply:
Yes, let me do that refactoring.